File: tests/fixtures/new_features.py
from stdio import *

# Test 1: list[T, n] for arrays
def test_list_arrays() -> void:
    a: list[int, 10]
    b: list[-char, 5]
    c: list[type[Node], 3]

# Test 2: Flexible array member list[T]
class Buffer:
    len: int
    data: list[char]

# Test 3: Bitfields bit[T, n]
class Flags:
    a: bit[unsigned[int], 3]
    b: bit[unsigned[int], 5]
    c: bit[int, 1]

# Test 4: Anonymous struct with @var
@var(point_a, point_b)
class _:
    x: int
    y: int

# Test 5: Anonymous union with @var
@var(data_u)
class _(Union):
    i: int
    f: float

# Test 6: Infinite loop for(;;)
def test_infinite_loop() -> void:
    i: int = 0
    while ():
        i ** _
        if i > 10:
            break

# Test 7: Do-while loop (still works)
def test_do_while() -> void:
    i: int = 0
    while ():
        printf("%d\n", i)
        i ** _
        if i < 5:
            continue

# Test 8: Preprocessor elif/else
if [DEBUG]:
    printf("debug mode\n")
elif [VERBOSE]:
    printf("verbose mode\n")
elif [QUIET]:
    printf("quiet mode\n")
else:
    printf("normal mode\n")

# Test 9: Preprocessor with not
if [not DISABLE_LOGGING]:
    printf("logging enabled\n")
elif [ERROR_ONLY]:
    printf("errors only\n")
else:
    printf("logging disabled\n")

# Test 10: match/case for switch with fallthrough
def test_switch(x: int) -> void:
    match x:
        case 1:
            printf("one\n")
            break
        case 2:
            printf("two or three\n")
        case 3:
            printf("three\n")
            break
        case _:
            printf("other\n")
            break

# Test 11: Combined example from design
class Node:
    data: int
    next: -type[Node]

@typedef(List)
class List:
    head: -type[Node]

def consume(l: -List) -> void:
    match l._.head._.data:
        case 0:
            break
        case _:
            break

def main() -> int:
    return 0

File: tests/golden_outputs/new_features.c
#include <stdio.h>
void test_list_arrays(void) {
    int a[10];
    char *b[5];
    struct Node c[3];
}
struct Buffer {
    int len;
    char data[];
};
struct Flags {
    unsigned int a : 3;
    unsigned int b : 5;
    int c : 1;
};
struct {
    int x;
    int y;
} point_a, point_b;
union {
    int i;
    float f;
} data_u;
void test_infinite_loop(void) {
    int i = 0;
    for (;;) {
        i++;
        if (i > 10) {
            break;
        }
    }
}
void test_do_while(void) {
    int i = 0;
    do {
        printf("%d\n", i);
        i++;
    } while (i < 5);
}
#ifdef DEBUG
printf("debug mode\n");
#elif defined(VERBOSE)
printf("verbose mode\n");
#elif defined(QUIET)
printf("quiet mode\n");
#else
printf("normal mode\n");
#endif
#ifndef DISABLE_LOGGING
printf("logging enabled\n");
#elif defined(ERROR_ONLY)
printf("errors only\n");
#else
printf("logging disabled\n");
#endif
void test_switch(int x) {
    switch (x) {
    case 1:
        printf("one\n");
        break;
    case 2:
        printf("two or three\n");
    case 3:
        printf("three\n");
        break;
    default:
        printf("other\n");
        break;
    }
}
struct Node {
    int data;
    struct Node *next;
};
typedef struct List {
    struct Node *head;
} List;
void consume(List *l) {
    switch (l->head->data) {
    case 0:
        break;
    default:
        break;
    }
}
int main(void) {
    return 0;
}